packages/jspsych/src/modules/randomization.ts

```diff
@@ -220,56 +216,43 @@ export function sampleWithReplacement(arr, size, weights) {
   return samp;
 }
 
-export function factorial(factors, repetitions, unpack) {
-  var factorNames = Object.keys(factors);
-
-  var factor_combinations = [];
-
-  for (var i = 0; i < factors[factorNames[0]].length; i++) {
-    factor_combinations.push({});
-    factor_combinations[i][factorNames[0]] = factors[factorNames[0]][i];
-  }
-
-  for (var i = 1; i < factorNames.length; i++) {
-    var toAdd = factors[factorNames[i]];
-    var n = factor_combinations.length;
-    for (var j = 0; j < n; j++) {
-      var base = factor_combinations[j];
-      for (var k = 0; k < toAdd.length; k++) {
-        var newpiece = {};
-        newpiece[factorNames[i]] = toAdd[k];
-        factor_combinations.push(Object.assign({}, base, newpiece));
+export function factorial(factors, repetitions = 1, unpack = false) {
+  let design = [{}];
+  for (const factor of Object.keys(factors)) {
+    let new_design = [];
+    for (const level of factors[factor]) {
+      for (const cell of design) {
+        let n = {};
+        n[factor] = level;
+        new_design.push(Object.assign({}, cell, n));
```

Comment:
> but it didn't play nice with the stuff below and my debugging attempts hit a (small) wall so I'm leaving it for now

Then here's the full version :wink: 
```suggestion
  for (const [factorName, factor] of Object.entries(factors)) {
    const new_design = [];
    for (const level of factor) {
      for (const cell of design) {
        new_design.push({ ...cell, factorName: level });
```

Reply:
So this looks very close to what I did, but I was getting an error on L223 that `factor` was of type `unknown` and was therefore not iterable. I'm guessing I made some upstream mistake, but at least I had the right logic in mind :grin:

Reply:
Oh, that's because we do not type the `factors` parameter... Maybe assign it a very general type already that we can specify further one day.